### PR TITLE
Add Buildkite Release Builds

### PR DESCRIPTION
**Pull request by jkmassel**
Adds to ability to produce release builds using Buildkite and the API. 

**To Test:**
Note that https://buildkite.com/wordpress-mobile/wordpress-ios/builds/4791 builds successfully, exiting with the error `This binary already exists`.

Best tested in combination with https://github.com/wordpress-mobile/release-toolkit/pull/333

### CODE REVIEW

File: .buildkite/release-builds.yml
Comment (mokagio) [inline .buildkite/release-builds.yml]: There might be some copy-paste leftover here, but I'm not sure at which level.

If this pipeline is meant to run on a schedule then I would suggest updating the explanation comment to something like:

```suggestion
# This script is run via Buildkite's scheduled jobs feature.
#
# It's meant to build and deploy various installable builds on a periodic
# async basis.

```

But, I checked the current schedules, and there's none:

image

That could just be because this PR hasn't been merged yet, and configuring before it's merged it schedule would result in failed builds. But it also made me wonder whether this pipeline is actually meant to run on a schedule at all? If that's not the case, then I'd simply remove the whole comment and trust that a reader would gather what this is about based on the content and name alone.

```suggestion
```

File: .buildkite/release-builds.yml
Comment (mokagio) [inline .buildkite/release-builds.yml]: What do you think of specifying this build goes to ASC, too?

```suggestion
  - label: "🛠 Jetpack Release Build (App Store Connect)"
```

One could argue that since there's only one kind of release build, people interested would know where to find it. But, since the other steps specify where the build is, it might be handy to do the same here, too.

File: fastlane/Fastfile
Comment (mokagio) [inline fastlane/Fastfile]: What's necessary to remove this?

It's used in the `build_and_upload_beta_release` and `build_and_upload_stable_release`: 

https://github.com/wordpress-mobile/WordPress-iOS/blob/7cda265ae9c04521d6e7d35a87e5464f3c244c94/fastlane/Fastfile#L488-L518

Which are in turn used by in CircleCI:

https://github.com/wordpress-mobile/WordPress-iOS/blob/7cda265ae9c04521d6e7d35a87e5464f3c244c94/.circleci/config.yml#L186-L202

So I'm guessing it's just a matter of waiting for the transition to Buildkite to be complete?

If that's the case, do you mind adding that information in the comment? Just in case we forget and someone lands on it in the future when CircleCI is gone already.

```suggestion
  # build_and_upload_release [Deprecated - Can be removed once CircleCI is no longer in use]
```

Comment (mokagio) [reply fastlane/Fastfile]: > So I'm guessing it's just a matter of waiting for the transition to Buildkite to be complete?

Actually, I just noticed the `build_and_upload_beta_release` lane is used in this PR, too, so there's probably more to it.

File: .buildkite/commands/installable-build-wordpress.sh
Comment (mokagio) [inline .buildkite/commands/installable-build-wordpress.sh]: I thought we'd want installable builds to go out via App Center? 

In case this is a typo:

```suggestion
bundle exec fastlane build_and_upload_app_center skip_confirm:true
```

But then, I wonder, what would be the difference between this command and the new `.buildkite/commands/release-build-wordpress-internal.sh`?

File: .buildkite/commands/release-build-wordpress.sh
Comment (mokagio) [inline .buildkite/commands/release-build-wordpress.sh]: 🤔 this doesn't use the new lane ASC. Is there a reason? 

Otherwise, I guess one of the following two suggestion would apply:

```suggestion
bundle exec fastlane build_and_upload_app_store_connect beta_release:true skip_confirm:true create_gh_release:true
```

```suggestion
bundle exec fastlane build_and_upload_app_store_connect skip_confirm:true create_gh_release:true
```

I find the word "beta" and "release" used in the same "sentence" a bit confusing though 🤔

File: .buildkite/commands/release-build-jetpack.sh
Comment (mokagio) [inline .buildkite/commands/release-build-jetpack.sh:19]: Have you considered extracting all this boilerplate in a file to `source`?

I think it would still work in Buidkite, but if it didn't, we could update the `command` node from calling this script only to first calling the new shared setup and then running the remaining code (below this comment).

I think duplicating the setup keeps each implementation flexible, but I wonder how much flexibility we need here?

Conversely, here we have two fixes that we'll need to address in the future (The Xcode 13.0 Ruby setup and the `brew update` call) which we'll need to replicate across all commands when the time comes. Granted, it's a simple, mechanical, one-time (🙏 ) operation, so it wouldn't be the end of the world. 😄

**Review by mokagio — COMMENTED** [review]
I left a bunch of questions.

Other than a few suggestions, this looks good so far, but I'm not sure what's the intent of this new pipeline. Run on a schedule? Replace the current installable builds setup? I don't think it's meant to run for the tags and replace the current CircleCI release workflow, otherwise it would be part of the main pipeline, not standalone. I fear some of my questions might be off point without that bit of information, sorry about that.

File: .buildkite/release-builds.yml
Comment (jkmassel) [reply .buildkite/release-builds.yml]: Yikes – just some major copy/paste blindness 🤦 

In this case it's meant to be called exclusively via the API, which I've noted in 45cf8a5505c7a8425d3101780d34fc68373e54e9

File: .buildkite/release-builds.yml
Comment (jkmassel) [reply .buildkite/release-builds.yml]: Yep this is a nice one!

File: .buildkite/commands/installable-build-wordpress.sh
Comment (jkmassel) [reply .buildkite/commands/installable-build-wordpress.sh]: Shoot – I rebased this PR and I think I messed up the merge pretty good. Sorry for the confusion – I've fixed it in edcfabc45bb3afbaa468a767c57af2c6be33d2a8.

File: .buildkite/commands/release-build-wordpress.sh
Comment (jkmassel) [reply .buildkite/commands/release-build-wordpress.sh]: Yep – because I'm bad at merges 🤦

File: .buildkite/commands/release-build-jetpack.sh
Comment (jkmassel) [reply .buildkite/commands/release-build-jetpack.sh:19]: This is a good question – I considered some DRY for this, but was worried that if these diverge for other reasons we might end up with a mess, hence keeping them separate from the start.

I think once we're able to clean out the `brew` and `install bundler` stuff, we might be able to do your idea of a `prepare_dependencies.sh` or something in the `command` block – that seems like a nice clean way to address this 🙂.

IMHO that's for a future PR though

**Comment by jkmassel** [comment]
You can test the <strong>Jetpack</strong> changes on this Pull Request by downloading it from AppCenter <a href='https://install.appcenter.ms/orgs/automattic/apps/jetpack-installable-builds/'>here</a> with build number: <code>pr17862-9680da1</code>. IPA is available <a href='https://appcenter-filemanagement-distrib1ede6f06e.azureedge.net/ace385b8-1fb3-4c18-ad2b-597bdfcbae2d/Jetpack%20Alpha.ipa?sv=2019-02-02&sr=c&sig=GUm4Cl9k%2BfDHCDjZcXJWHoNEb5u6IG%2B6GVtE3PNNEjc%3D&se=2022-02-03T21%3A25%3A24Z&sp=r'>here</a>. If you need access to this, you can ask a maintainer to add you.

**Comment by jkmassel** [comment]
You can test the <strong>WordPress</strong> changes on this Pull Request by downloading it from AppCenter <a href='https://install.appcenter.ms/orgs/automattic/apps/WPiOS-One-Offs/'>here</a> with build number: <code>pr17862-9680da1</code>. IPA is available <a href='https://appcenter-filemanagement-distrib5ede6f06e.azureedge.net/b7658e35-07af-4e5d-95ff-7b86b9427606/WordPress%20Alpha.ipa?sv=2019-02-02&sr=c&sig=ftDF0fAWFKUz8ThNZS2MVUsaVcOP9U2n2M4TWRGrdJM%3D&se=2022-02-03T21%3A26%3A55Z&sp=r'>here</a>. If you need access to this, you can ask a maintainer to add you.

File: .buildkite/commands/release-build-jetpack.sh
Comment (mokagio) [reply .buildkite/commands/release-build-jetpack.sh:19]: > IMHO that's for a future PR though

+1, I'm all for incremental improvements and small steps.

**Review by mokagio — APPROVED** [review]
Thank you for addressing my questions 😄 

After your edits, the pipelines and commands look much more straightforward 👍 

I used the new release toolkit action to start [a build](https://buildkite.com/automattic/wordpress-ios/builds/4849#_) of this new pipeline and the two ASC steps failed as expected with the `ERROR ITMS-90189: Redundant Binary Upload` error.

image

:shipit: 

Very excited to get to use this for 19.2 💙 If you don't get to finish wiring the bits by calling the new release toolkit action, I'll do it as part of the code freeze.

**Pull request merged**
